File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int maxWater = 0; // Initialize maximum area to 0
        int lp = 0, rp = height.length - 1; // Left and right pointers

        while (lp < rp) {
            int width = rp - lp;
            int ht = Math.min(height[lp], height[rp]);
            int currWater = width * ht; // Calculate the current area
            maxWater = Math.max(maxWater, currWater); // Update maxWater if current area is larger

            // Move the pointer corresponding to the smaller height
            if (height[lp] < height[rp]) {
                lp++;
            } else {
                rp--;
            }
        }

        return maxWater; // Return the maximum area found
    }
}

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {  // TC -> O(n)
        int maxWater = 0; // ans
        int lp = 0, rp = height.size() - 1;

        while (lp < rp) {
            int w = rp - lp;
            int ht = min(height[lp], height[rp]);
            int currWater = w * ht;
            maxWater = max(maxWater, currWater);

            height[lp] < height[rp] ? lp++ : rp--;
        }
        return maxWater;
    }
};